concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -696,26 +778,266 @@ handleUpdateContract wtc uAmount uAddress uReceiveName uMessage =
   where senderAccount = wtc ^. wtcSenderAccount
         meta = wtc ^. wtcTransactionHeader
         senderAddress = thSender meta
+        checkAndGetBalanceV1 = checkAndGetBalanceAccountV1 senderAddress senderAccount
+        checkAndGetBalanceV0 = checkAndGetBalanceAccountV0 senderAddress senderAccount
         c = do
-          ins <- getCurrentContractInstanceTicking uAddress
-          -- Now invoke the general handler for contract messages.
-          handleMessage senderAddress
-                        ins
-                        (Right (senderAddress, senderAccount))
-                        uAmount
-                        uReceiveName
-                        uMessage
-
+          getCurrentContractInstanceTicking uAddress >>= \case
+            InstanceV0 ins -> 
+                -- Now invoke the general handler for contract messages.
+                handleContractUpdateV0 senderAddress
+                              ins
+                              checkAndGetBalanceV0
+                              uAmount
+                              uReceiveName
+                              uMessage
+            InstanceV1 ins -> do
+              handleContractUpdateV1 senderAddress ins checkAndGetBalanceV1 uAmount uReceiveName uMessage >>= \case
+                Left cer -> rejectTransaction (WasmV1.cerToRejectReasonReceive uAddress uReceiveName uMessage cer)
+                Right (_, events) -> return (reverse events)
+
+-- |Check that the account has sufficient balance, and construct credentials of the account.
+checkAndGetBalanceAccountV1 :: (TransactionMonad pv m, AccountOperations m)
+    => AccountAddress -- ^Used address
+    -> IndexedAccount m
+    -> Amount
+    -> m (Either WasmV1.ContractCallFailure (Address, [ID.AccountCredential], (Either ContractAddress IndexedAccountAddress)))
```

Comment:
As opposed to? This is the most precise type we have. If it was rejectreason we'd have problems reporting it back to the contract since there are many cases which cannot happen.

Reply:
I was thinking of the regular account -> account transfer. Never mind 😊